library/x509write_csr.c

```diff
@@ -273,24 +280,33 @@ int mbedtls_x509write_csr_pem( mbedtls_x509write_csr *ctx, unsigned char *buf, s
                        int (*f_rng)(void *, unsigned char *, size_t),
                        void *p_rng )
 {
-    int ret;
-    unsigned char output_buf[4096];
+    int ret = 0;
+    unsigned char *output_buf;
```

Comment:
`mbedtls_pem_write_buffer` can work in place, so we could use `buf` as the temporary buffer to build the DER. `output_buf` is not actually needed.

There is one edge case where this doesn't work, which is if `buf` points to shared memory that can't be trusted to yield back what was last written. However, Mbed TLS does not in general advertise that this is possible. It just happens to be possible with this particular function at this particular stage of the evolution of the library. So we don't need to preserve this property.

Reply:
If `mbedtls_pem_write_buffer()` is documented as being able to work in place, that's a good direction to go. I don't like how this function is implemented very much currently.

Reply:
It's documented but not tested. Looking at the implementation, I'm not sure if it works for arbitrary overlap, even when the PEM starts before the DER.

Reply:
`
 * \note            This function may be called with overlapping \p der_data
 *                  and \p buf buffers.
`

Good enough for me.